core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -667,6 +710,123 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+  /**
+   * Function to return an average; if the count is 0, so is the average.
+   * @param value value to average
+   * @param count event count to divide by
+   * @return the average, or 0 if the counter is itself 0
+   */
+  private def average(value: Long, count: Long): Long = {
+    if (count> 0) value / count else 0
+  }
+
+  override val sourceName = "history.fs"
+
+  private val name = MetricRegistry.name(sourceName)
+
+  /** Number of updates. */
+  val updateCount = new Counter()
+
+  /** Number of update failures. */
+  val updateFailureCount = new Counter()
+
+  /** Number of events replayed as listing merge. */
+  val historyEventCount = new Counter()
+
+  /** Timer of listing merges. */
+  val historyMergeTimer = new Timer()
+
+  /** Total time to merge all histories. */
+  val historyTotalMergeTime = new Counter()
+
+  /** Average time to load a single event in the App UI */
+  val historyEventMergeTime = new LambdaLongGauge(() =>
+    average(historyTotalMergeTime.getCount, historyEventCount.getCount))
+
+  /** Number of events replayed. */
+  val appUIEventCount = new Counter()
+
+  /** Update duration timer. */
+  val updateTimer = new Timer()
+
+  /** Time the last update was attempted. */
+  val updateLastAttempted = new Timestamp()
+
+  /** Time the last update succeded. */
+  val updateLastSucceeded = new Timestamp()
+
+  /** Number of App UI load operations. */
+  val appUILoadCount = new Counter()
+
+  /** Number of App UI load operations that failed due to a load/parse/replay problem. */
+  val appUILoadFailureCount = new Counter()
+
+  /** Number of App UI load operations that failed due to an unknown file. */
+  val appUILoadNotFoundCount = new Counter()
+
+  /** Statistics on time to load app UIs. */
+  val appUiLoadTimer = new Timer()
+
+  /** Total load time of all App UIs. */
+  val appUITotalLoadTime = new Counter()
+
+  /** Average time to load a single event in the App UI */
+  val appUIEventReplayTime = new LambdaLongGauge(() =>
+    average(appUITotalLoadTime.getCount, appUIEventCount.getCount))
+
+  private val countersAndGauges = Seq(
+    ("history.merge.event.count", historyEventCount),
+    ("history.merge.event.time", historyEventMergeTime),
+    ("history.merge.duration", historyTotalMergeTime),
+    ("update.count", updateCount),
+    ("update.failure.count", updateFailureCount),
+    ("update.last.attempted", updateLastAttempted),
+    ("update.last.succeeded", updateLastSucceeded),
+    ("appui.load.count", appUILoadCount),
+    ("appui.load.duration", appUITotalLoadTime),
+    ("appui.load.failure.count", appUILoadFailureCount),
+    ("appui.load.not-found.count", appUILoadNotFoundCount),
+    ("appui.event.count", appUIEventCount),
+    ("appui.event.replay.time", appUIEventReplayTime)
+  )
+
+  private val timers = Seq (
+    ("update.timer", updateTimer),
+    ("history.merge.timer", historyMergeTimer),
+    ("appui.load.timer", appUiLoadTimer))
+
+  val allMetrics = countersAndGauges ++ timers
```

Comment:
You don't seem to need these lists outside of the context of calling `register`, so can you instead inline the list into the call to avoid having these fields?